src/event/sys/windows/parse.rs

```diff
@@ -376,3 +373,42 @@ fn parse_mouse_event_record(
         modifiers,
     }))
 }
+
+const HIGH_SURROGATE_FIRST: u16 = 0xD800;
+const LOW_SURROGATE_FIRST: u16 = 0xDC00;
+const LOW_SURROGATE_LAST: u16 = 0xDFFF;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_handle_surrogate() {
+        fn testcase(mut buf: Option<u16>, new: u16) -> (Option<u16>, Option<char>) {
+            let ch = handle_surrogate(&mut buf, new);
+            (buf, ch)
+        }
+        assert_eq!(
+            [
+                // Continuous high surrogates
+                testcase(Some(0xD800), 0xDBFF),
+                // A normal surrogate pair
+                testcase(Some(0xD800), 0xDC00),
+                // An empty buffer and a high surrogate
+                testcase(None, 0xDBFF),
+                // An empty buffer and a low surrogate
+                testcase(None, 0xDC00),
+            ],
+            [
+                // The newer one will be stored
+                (Some(0xDBFF), None),
+                // A valid char will return
+                (None, Some('\u{10000}')),
+                // The high surrogate will be stored
+                (Some(0xDBFF), None),
+                // The low surrogate will be ignored
+                (None, None),
+            ]
+        );
```

Comment:
Simplify (individual assertions, no testcase function, descriptions on the assertions, arranged in order that makes most sense (empty to full)):
```suggestion
        let mut buf = None;
        let next = handle_surrogate(&mut buf, 0xDC00);
        assert_eq!(
            (buf, next),
            (None, None),
            "Unexpected low surrogate is ignored"
        );

        let mut buf = None;
        let next = handle_surrogate(&mut buf, 0xDBFF);
        assert_eq!(
            (buf, next),
            (Some(0xDBFF), None),
            "High surrogate is buffered"
        );

        let mut buf = Some(0xD800);
        let next = handle_surrogate(&mut buf, 0xDBFF);
        assert_eq!(
            (buf, next),
            (Some(0xDBFF), None),
            "Consecutive high surrogates"
        );

        let mut buf = Some(0xD800);
        let next = handle_surrogate(&mut buf, 0xDC00);
        assert_eq!(
            (buf, next),
            (None, Some('\u{10000}')),
            "Valid Surrogate pair"
        );
```